File: src/audio/audio.cc
#include "src/audio/audio.h"

#include <vector>

#include "src/framework/log.h"
#include "src/framework/time.h"

namespace Looper
{

/**
 * Whether portaudio has been initialized.
 */
bool is_init = false;

bool init_audio()
{
    /*
     * Initialize portaudio if it hasn't been initialized yet.
     */
    auto err = Pa_Initialize();
    ASSERT(err == paNoError,
           "Failed to initialize PortAudio: %s",
           Pa_GetErrorText(err));
    is_init = true;
    return true;
}

/**
 * Get the device index matching the input name.
 *
 * @param[out] index The matching device index.
 * @param name The name of the device.
 * @param is_input Whether the device is an input.
 * @param is_output Whether the device is an output.
 */
bool get_device_index(PaDeviceIndex &index,
                      const std::string &name,
                      bool is_input,
                      bool is_output)
{
    ASSERT(is_init, "PortAudio was not initialized. Call init_audio()");

    /*
     * Get a list of devices.
     */
    const PaDeviceIndex numDevices = Pa_GetDeviceCount();
    ASSERT(numDevices >= 0,
           "Failed to get devices: %s",
           Pa_GetErrorText(numDevices));

    /*
     * Search for the device in the list of devices.
     */
    std::vector<std::string> available_devices;
    for (PaDeviceIndex i = 0; i < numDevices; i++)
    {
        /*
         * Return the first match.
         */
        const PaDeviceInfo *device_info = Pa_GetDeviceInfo(i);
        const std::string device_name = device_info->name;
        if (device_name.find(name) != std::string::npos)
        {
            if (is_input && device_info->maxInputChannels < 1)
            {
                LOG(INFO, "Device %s was not an input", device_info->name);
                continue;
            }
            else if (is_output && device_info->maxOutputChannels < 1)
            {
                LOG(INFO, "Device %s was not an output", device_info->name);
                continue;
            }

            index = i;
            return true;
        }

        /*
         * Keep track of this device in case we don't find the one we want.
         */
        available_devices.push_back(device_info->name);
    }

    /*
     * If we reach this point, we didn't find a device. Print out all of the
     * devices we found to help a user debug.
     */
    LOG(ERROR, "Could not find audio device \"%s\"", name.c_str());
    LOG(ERROR, "Available devices:");
    for (const std::string &device : available_devices)
    {
        LOG(ERROR, "  - %s", device.c_str());
    }

    return false;
}

/**
 * Callback for when we receive audio.
 *
 * @see PaStreamCallback
 */
int audio_input_callback(const void *input_buffer,
                         void *output_buffer,
                         unsigned long frames_per_buffer,
                         const PaStreamCallbackTimeInfo *time_info,
                         PaStreamCallbackFlags status_flags,
                         void *data)
{
    InputDevice *device = (InputDevice *)data;
    if (frames_per_buffer != device->buf.size())
    {
        LOG(ERROR, "Portaudio gave an unusual number of frames.");
        return paAbort;
    }

    /*
     * Wait until the previous data has been processed.
     */
    std::unique_lock<std::mutex> lock(device->mutex);
    while (device->buffer_full)
    {
        device->cv.wait(lock);
    }

    /*
     * Copy off the device buffer to our internal buffer.
     */
    float *in = (float *)input_buffer;
    for (size_t i = 0; i < device->buf.size(); i++)
    {
        device->buf[i] = *in;
        in++;
    }

    device->buffer_full = true;
    device->cv.notify_one();

    return paContinue;
}

/**
 * Callback for when we need to send audio.
 *
 * @see PaStreamCallback
 */
int audio_output_callback(const void *input_buffer,
                          void *output_buffer,
                          unsigned long frames_per_buffer,
                          const PaStreamCallbackTimeInfo *time_info,
                          PaStreamCallbackFlags status_flags,
                          void *data)
{
    OutputDevice *device = (OutputDevice *)data;
    if (frames_per_buffer != device->buf.size())
    {
        LOG(ERROR, "Portaudio gave an unusual number of frames.");
        return paAbort;
    }

    /*
     * Wait until the underlying stream is ready to send more data.
     */
    std::unique_lock<std::mutex> lock(device->mutex);
    while (!device->buffer_full)
    {
        device->cv.wait(lock);
    }

    /*
     * Copy off the device buffer to our internal buffer.
     */
    float *out = (float *)output_buffer;
    for (size_t i = 0; i < device->buf.size(); i++)
    {
        *out = device->buf[i];
        out++;
    }

    /*
     * Notify everyone that the buffer is empty.
     */
    device->buffer_full = false;
    device->cv.notify_one();

    return paContinue;
}

InputDevice::InputDevice(const BlockConfig _configs) : Source(_configs) {}

bool InputDevice::init()
{
    std::string audio_device_name;
    ASSERT(configs.get_string("device", audio_device_name),
           "Missing audio device name");

    /*
     * Get the device.
     */
    PaDeviceIndex index = -1;
    ASSERT(get_device_index(index, audio_device_name, true, false),
           "Error finding device.");
    const PaDeviceInfo *info = Pa_GetDeviceInfo(index);

    /*
     * Parameters.
     */
    const PaStreamParameters params = {
        .channelCount = 1,
        .device = index,
        .sampleFormat = SAMPLE_WIDTH,
        .suggestedLatency = info->defaultLowInputLatency,
        .hostApiSpecificStreamInfo = NULL};

    /*
     * Open an audio I/O stream.
     */
    PaError err = Pa_OpenStream(&pa_stream,
                                &params,
                                NULL,
                                SAMPLE_RATE,
                                SAMPLES_PER_BUFFER,
                                paNoFlag,
                                &audio_input_callback,
                                (void *)this);
    ASSERT(err == paNoError,
           "Failed to open input device: %s: %s",
           audio_device_name.c_str(),
           Pa_GetErrorText(err));

    /*
     * Read back the stream info.
     */
    const PaStreamInfo *stream_info = Pa_GetStreamInfo(pa_stream);
    LOG(DEBUG,
        "Input stream info: sample_rate=%f, latency=%f",
        stream_info->sampleRate,
        stream_info->inputLatency);

    err = Pa_StartStream(pa_stream);
    ASSERT(err == paNoError,
           "Failed to start device stream: %s: %s",
           audio_device_name.c_str(),
           Pa_GetErrorText(err));

    return true;
}

bool InputDevice::read()
{
    /*
     * Simple copy.
     */
    std::unique_lock<std::mutex> lock(mutex);
    while (!buffer_full)
    {
        cv.wait(lock);
    }

    std::copy(std::begin(buf), std::end(buf), std::begin(*stream));

    buffer_full = false;
    cv.notify_all();

    return true;
}

OutputDevice::OutputDevice(const BlockConfig _configs)
    : Sink(_configs), buffer_full(false)
{
}

bool OutputDevice::init()
{
    std::string audio_device_name;
    ASSERT(configs.get_string("device", audio_device_name),
           "Missing audio device name");

    /*
     * Get the device.
     */
    PaDeviceIndex index = -1;
    ASSERT(get_device_index(index, audio_device_name, false, true),
           "Error finding device.");
    const PaDeviceInfo *info = Pa_GetDeviceInfo(index);
    LOG(DEBUG, "API: %s", Pa_GetHostApiInfo(info->hostApi)->name);

    /*
     * Parameters.
     */
    const PaStreamParameters params = {
        .channelCount = 1,
        .device = index,
        .sampleFormat = SAMPLE_WIDTH,
        .suggestedLatency = info->defaultLowOutputLatency,
        .hostApiSpecificStreamInfo = NULL};

    /*
     * Open an audio I/O stream.
     */
    PaError err = Pa_OpenStream(&pa_stream,
                                NULL,
                                &params,
                                SAMPLE_RATE,
                                SAMPLES_PER_BUFFER,
                                paNoFlag,
                                &audio_output_callback,
                                (void *)this);
    ASSERT(err == paNoError,
           "Failed to open output device: %s: %s",
           audio_device_name.c_str(),
           Pa_GetErrorText(err));

    /*
     * Read back the stream info.
     */
    const PaStreamInfo *stream_info = Pa_GetStreamInfo(pa_stream);
    LOG(DEBUG,
        "Output stream info: sample_rate=%f, latency=%f",
        stream_info->sampleRate,
        stream_info->outputLatency);

    err = Pa_StartStream(pa_stream);
    ASSERT(err == paNoError,
           "Failed to start device stream: %s: %s",
           audio_device_name.c_str(),
           Pa_GetErrorText(err));

    return true;
}

bool OutputDevice::write()
{
    /*
     * Wait until the underlying stream is ready to receive more data.
     */
    std::unique_lock<std::mutex> lock(mutex);
    while (buffer_full)
    {
        cv.wait(lock);
    }

    /*
     * Simple copy.
     */
    std::copy(std::begin(*stream), std::end(*stream), std::begin(buf));

    /*
     * Notify the stream that there's more data available.
     */
    buffer_full = true;
    cv.notify_all();
    return true;
}

}  // namespace Looper

File: src/audio/audio.rs
//! Audio I/O Blocks.
//!
//! AudioSource [Source]:
//!     Required parameters:
//!         name: Anything
//!         type: "AudioSource"
//!         device: The name of the audio device on your computer. Leave empty and the resulting
//!                 error message will print a list of available devices.
//!         output_channel: The output channel.
//!
//! AudioSink [Sink]:
//!     Required parameters:
//!         name: Anything
//!         type: "AudioSink"
//!         device: The name of the audio device on your computer. Leave empty and the resulting
//!                 error message will print a list of available devices.
//!         input_channel: The output channel.

extern crate portaudio;

extern crate block;
extern crate config;
extern crate stream;
extern crate tempo;

use portaudio::{stream_flags::Flags, Blocking, DeviceIndex, DeviceInfo, Flow, PortAudio};

type PaStreamInput = portaudio::Input<stream::Sample>;
type PaStreamOutput = portaudio::Output<stream::Sample>;

/// Get a new PortAudio instance.
pub fn pa_get() -> Result<PortAudio, ()> {
    let pa = log::unwrap_abort!(PortAudio::new());
    println!("PortAudio initialized.");

    Ok(pa)
}

/// Get the audio device with the associated name.
fn get_device_index<'a>(
    pa: &'a PortAudio,
    name: &str,
    is_input: bool,
    is_output: bool,
) -> Result<(DeviceIndex, DeviceInfo<'a>), ()> {
    for device in log::unwrap_abort!(pa.devices()) {
        let (device_id, device_info) = log::unwrap_abort!(device);

        // Found the matching device -- assert that it meets our criteria.
        // If it doesn't, that's ok, let's try the next one.
        if device_info.name.contains(name) {
            if is_input && device_info.max_input_channels < 1 {
                println!("Not an input device! {}", device_info.name);
                continue;
            }
            if is_output && device_info.max_output_channels < 1 {
                println!("Not an output device! {}", device_info.name);
                continue;
            }

            return Ok((device_id, device_info));
        }
    }

    // If we reach this point, we didn't find a device.
    println!("Could not find audio device: \"{}\"", name);
    println!("Available devices:");
    for device in log::unwrap_abort!(pa.devices()) {
        let (_, device_info) = log::unwrap_abort!(device);
        println!("  - \"{}\"", device_info.name);
    }

    Err(())
}

// Audio Blocks

/// A valid Source Block which derives input from system audio hardware. This could be an external
/// USB microphone, an audio jack, or a built-in microphone.
pub struct AudioSource {
    /// The block name.
    name: String,

    /// The input [portaudio] stream.
    pa_stream: portaudio::Stream<Blocking<<PaStreamInput as Flow>::Buffer>, PaStreamInput>,

    /// The output [looper] stream buffer.
    stream: stream::Stream,
}

/// A valid Sink Block which writes a stream to a hardware speaker.
pub struct AudioSink {
    /// The block name.
    name: String,

    /// The output [portaudio] stream.
    pa_stream: portaudio::Stream<Blocking<<PaStreamOutput as Flow>::Buffer>, PaStreamOutput>,

    /// The input [looper] stream buffer.
    stream: stream::Stream,
}

impl AudioSource {
    /// Initialize a PortAudio AudioSource block.
    pub fn new(
        config: &config::BlockConfig,
        stream_catalog: &mut stream::StreamCatalog,
        pa: &PortAudio,
    ) -> Result<Self, ()> {
        // Read in parameters.
        let output_stream = config.get_str("output_channel")?;
        let device_name = config.get_str("device")?;

        // Load streams.
        let stream = stream_catalog.create_source(output_stream)?;

        // PortAudio wizardry follows...

        // Get the audio device.
        let (device_index, device_info) =
            log::unwrap_abort!(get_device_index(pa, device_name, true, false));

        // PortAudio stream parameters.
        let input_params = portaudio::StreamParameters::<stream::Sample>::new(
            device_index,
            1,    /* channel count */
            true, /* interleaved */
            device_info.default_low_output_latency,
        );

        // Open the audio stream.
        let settings = portaudio::InputStreamSettings {
            params: input_params,
            sample_rate: 44100.0,
            frames_per_buffer: stream::SAMPLES_PER_BUFFER as u32,
            flags: Flags::NO_FLAG,
        };
        let mut pa_stream = log::unwrap_abort_msg!(
            pa.open_blocking_stream(settings),
            format!("Error opening audio channel \"{}\"", device_name)
        );
        log::unwrap_abort!(pa_stream.start());

        Ok(AudioSource {
            name: config.name.to_owned(),
            pa_stream: pa_stream,
            stream: stream,
        })
    }

    /// Block until some number of samples is available to read.
    fn get_available_samples(&self) -> usize {
        match self.pa_stream.read_available() {
            Ok(available) => match available {
                portaudio::StreamAvailable::Frames(frames) => return frames as usize,
                _ => {
                    println!("Stream {} has overflowed", self.name);
                    return 0;
                }
            },
            Err(e) => {
                println!("Input stream error!! {}", e.to_string());
                return 0;
            }
        };
    }
}

impl AudioSink {
    /// Initialize a PortAudio AudioSink block.
    pub fn new(
        config: &config::BlockConfig,
        stream_catalog: &stream::StreamCatalog,
        pa: &PortAudio,
    ) -> Result<Self, ()> {
        // Read in parameters.
        let input_stream = config.get_str("input_channel")?;
        let device_name = config.get_str("device")?;

        // Load streams.
        let stream = stream_catalog.bind_sink(input_stream)?;

        // PortAudio wizardry follows...

        // Get the audio device.
        let (device_index, device_info) =
            log::unwrap_abort!(get_device_index(pa, device_name, false, true));

        // PortAudio stream parameters.
        let output_params = portaudio::StreamParameters::<stream::Sample>::new(
            device_index,
            1,    /* channel count */
            true, /* interleaved */
            device_info.default_low_output_latency,
        );

        // Open the audio stream.
        let settings = portaudio::OutputStreamSettings {
            params: output_params,
            sample_rate: 44100.0,
            frames_per_buffer: stream::SAMPLES_PER_BUFFER as u32,
            flags: Flags::NO_FLAG,
        };
        let mut pa_stream = log::unwrap_abort_msg!(
            pa.open_blocking_stream(settings),
            format!("Error opening audio channel \"{}\"", device_name)
        );
        log::unwrap_abort!(pa_stream.start());

        Ok(AudioSink {
            name: config.name.to_owned(),
            pa_stream: pa_stream,
            stream: stream,
        })
    }

    /// Block until some number of samples is available to write.
    fn get_available_samples(&self) -> usize {
        match self.pa_stream.write_available() {
            Ok(available) => match available {
                portaudio::StreamAvailable::Frames(frames) => return frames as usize,
                _ => {
                    println!("Stream {} has underflowed", self.name);
                    return 0;
                }
            },
            Err(e) => {
                println!("Output stream error!! {}", e.to_string());
                return 0;
            }
        };
    }
}

impl block::Source for AudioSource {
    fn read(&mut self, _: &block::PlaybackState) {
        let mut stream = (*self.stream).borrow_mut();

        let mut index: usize = 0;
        while index < stream::SAMPLES_PER_BUFFER {
            // Get the number of available samples and read them in from PA.
            let samples_remaining = stream::SAMPLES_PER_BUFFER - index;
            let num_samples = std::cmp::min(self.get_available_samples(), samples_remaining);

            // Read samples from the buffer.
            match self.pa_stream.read(num_samples as u32) {
                Ok(samples) => {
                    for i in 0..samples.len() {
                        stream[i + index] = samples[i];
                    }
                }
                Err(e) => {
                    println!("Failed to read audio input: {}", e.to_string());
                    break;
                }
            };

            index += num_samples;
        }
    }
}

impl block::Sink for AudioSink {
    fn write(&mut self, _: &block::PlaybackState) {
        let stream = (*self.stream).borrow();

        let mut index: usize = 0;
        while index < stream::SAMPLES_PER_BUFFER {
            // Get the number of available samples and read them in from PA.
            let samples_remaining = stream::SAMPLES_PER_BUFFER - index;
            let num_samples = std::cmp::min(self.get_available_samples(), samples_remaining);

            // Write samples out to the buffer.
            match self.pa_stream.write(num_samples as u32, |output| {
                for i in 0..num_samples {
                    output[i] = stream[i + index];
                }
            }) {
                Ok(_) => (),
                Err(e) => {
                    println!("Failed to write audio output: {}", e.to_string());
                    break;
                }
            }

            index += num_samples;
        }
    }
}
